lib/IRGen/StackSizeLimitTransform.cpp

```diff
@@ -0,0 +1,307 @@
+//===--- StackSizeLimitTransform.cpp --------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// This is a transform that ensures that swift allocations never blow out the
+/// stack. We do this by limiting per function the amount of stack space
+/// allocated and otherwise, heap allocate classes, values.
+///
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-irgen-stack-size-limit"
+
+#include "ClassTypeInfo.h"
+#include "IRGen.h"
+#include "IRGenModule.h"
+#include "swift/AST/IRGenOptions.h"
+#include "swift/Basic/BlotMapVector.h"
+#include "swift/Basic/FrozenMultiMap.h"
+#include "swift/IRGen/IRGenSILPasses.h"
+#include "swift/SIL/InstructionUtils.h"
+#include "swift/SIL/SILBasicBlock.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILInstructionWorklist.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/CanonicalizeInstruction.h"
+#include "swift/SILOptimizer/Utils/InstOptUtils.h"
+#include "llvm/IR/Constants.h"
+
+using namespace swift;
+using namespace swift::irgen;
+
+//===----------------------------------------------------------------------===//
+//                                  Utility
+//===----------------------------------------------------------------------===//
+
+/// Given an alloc_stack \p asi and the next element to process, convert asi's
+/// users to use a new heap allocated box without invalidating next.
+static SILBasicBlock::iterator
+convertStackBoxToHeapBox(AllocStackInst *asi, SILBasicBlock::iterator next) {
+  SILBuilderWithScope builder(asi);
+  auto *box =
+      builder.createAllocBox(asi->getLoc(), asi->getType().getObjectType());
+  auto *projectBox = builder.createProjectBox(asi->getLoc(), box, 0);
+  while (!asi->use_empty()) {
+    auto *use = *asi->use_begin();
+
+    // Replace all of the alloc_stack's dealloc_stack with a destroy_value of
+    // the box.
+    if (auto *dsi = dyn_cast<DeallocStackInst>(use->getUser())) {
```

Comment:
This looks like it's ending up with a double destroy of the payload: the original `alloc_stack` has a `destroy_addr` _and_ a `dealloc_stack` as use. Doesn't the box destructor destroy the payload?

Reply:
You are correct. I would need to dealloc_partial_ref so we treat the box as uninitialized. You are correct.